--- a/17_shader_vertexbuffer.vert
+++ b/21_shader_uniformbuffer.vert
@@ -1,6 +1,12 @@
 #version 450
 #extension GL_ARB_separate_shader_objects : enable
 
+layout(binding = 0) uniform UniformBufferObject {
+    mat4 model;
+    mat4 view;
+    mat4 proj;
+} ubo;
+
 // NOTE: names must match the `Vertex` struct in Rust
 layout(location = 0) in vec2 pos;
 layout(location = 1) in vec3 color;
@@ -12,6 +18,6 @@ out gl_PerVertex {
 };
 
 void main() {
-    gl_Position = vec4(pos, 0.0, 1.0);
+    gl_Position = ubo.proj * ubo.view * ubo.model * vec4(pos, 0.0, 1.0);
     fragColor = color;
-}
+}
\ No newline at end of file
